File: backend/src/main/java/edu/cit/mantilla/inventorypro/service/CategoryService.java
package edu.cit.mantilla.inventorypro.service;

import edu.cit.mantilla.inventorypro.entity.Category;
import edu.cit.mantilla.inventorypro.repository.CategoryRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;

@Service
public class CategoryService {

    @Autowired
    private CategoryRepository categoryRepository;

    /**
     * Get all active categories (non-deleted)
     * @return List of active categories
     */
    public List<Category> getAllActiveCategories() {
        return categoryRepository.findAllByDeletedAtIsNull();
    }

    /**
     * Get all categories including deleted ones (for reports)
     * @return List of all categories
     */
    public List<Category> getAllCategories() {
        return categoryRepository.findAll();
    }

    /**
     * Create a new category
     * @param category Category to create
     * @return Created category
     */
    public Category createCategory(Category category) {
        return categoryRepository.save(category);
    }

    /**
     * Update an existing category
     * @param id Category ID
     * @param updated Updated category data
     * @return Updated category
     * @throws RuntimeException if category not found
     */
    public Category updateCategory(Long id, Category updated) {
        Category existing = categoryRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Category not found with id: " + id));
        
        existing.setName(updated.getName());
        existing.setDescription(updated.getDescription());
        
        return categoryRepository.save(existing);
    }

    /**
     * Soft delete a category by setting deletedAt timestamp
     * @param id Category ID to soft delete
     * @throws RuntimeException if category not found
     */
    public void softDeleteCategory(Long id) {
        Category existing = categoryRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Category not found with id: " + id));
        
        existing.setDeletedAt(LocalDateTime.now());
        categoryRepository.save(existing);
    }
}

File: web/frontend/src/api/services/categoryService.js
/**
 * Category Service
 * Handles all category-related API calls
 * Uses API Client Factory - no boilerplate code
 */

import { apiClient } from '../apiClient';
import { apiConfig } from '../apiConfig';

export const categoryService = {
  /**
   * Get all active categories (non-deleted)
   */
  async getCategories() {
    return apiClient.get(apiConfig.ENDPOINTS.CATEGORIES.LIST);
  },

  /**
   * Get all categories including soft-deleted (for reports)
   */
  async getAllCategories() {
    return apiClient.get(apiConfig.ENDPOINTS.CATEGORIES.ALL);
  },

  /**
   * Create new category
   */
  async createCategory(data) {
    return apiClient.post(apiConfig.ENDPOINTS.CATEGORIES.CREATE, data);
  },

  /**
   * Update category
   */
  async updateCategory(id, data) {
    const endpoint = apiConfig.ENDPOINTS.CATEGORIES.UPDATE.replace(':id', id);
    return apiClient.put(endpoint, data);
  },

  /**
   * Delete category (soft delete)
   */
  async deleteCategory(id) {
    const endpoint = apiConfig.ENDPOINTS.CATEGORIES.DELETE.replace(':id', id);
    return apiClient.delete(endpoint);
  },
};
